File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/promisehandlers/ReplayCurrentTupleHandler.scala
package edu.uci.ics.amber.engine.architecture.pythonworker.promisehandlers

import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand

object ReplayCurrentTupleHandler {
  final case class ReplayCurrentTuple() extends ControlCommand[Unit]
}

File: core/amber/src/main/python/core/architecture/handlers/replay_current_tuple_handler.py
import itertools

from core.architecture.handlers.handler_base import Handler
from core.architecture.managers.context import Context
from proto.edu.uci.ics.amber.engine.architecture.worker import (
    ReplayCurrentTupleV2,
    WorkerState,
)


class ReplayCurrentTupleHandler(Handler):
    cmd = ReplayCurrentTupleV2

    def __call__(self, context: Context, command: cmd, *args, **kwargs):
        if not context.state_manager.confirm_state(WorkerState.COMPLETED):
            # chain the current input tuple back on top of the current iterator to
            # be processed once more
            context.tuple_processing_manager.current_input_tuple_iter = itertools.chain(
                [context.tuple_processing_manager.current_input_tuple],
                context.tuple_processing_manager.current_input_tuple_iter,
            )
        return None
